velox/exec/GroupingSet.cpp

```diff
@@ -412,4 +444,245 @@ const HashLookup& GroupingSet::hashLookup() const {
   return *lookup_;
 }
 
+namespace {
+bool maybeReserve(int64_t increment, memory::MemoryUsageTracker& tracker) {
+  constexpr int32_t kGrowthQuantum = 8 << 20;
+  auto addedReservation = bits::roundUp(increment, kGrowthQuantum);
+  // We look up the tracker tree to see if there is a parent that could have
+  // space. If some parent could have space we take the chance and try to
+  // increase reservation.
+  auto candidate = &tracker;
+  while (candidate) {
+    auto limit = candidate->maxTotalBytes();
+    if (limit == memory::kMaxMemory && candidate->parent()) {
+      candidate = candidate->parent();
+      continue;
+    }
+    if (limit - candidate->getCurrentTotalBytes() > addedReservation) {
+      try {
+        tracker.reserve(addedReservation);
+      } catch (const std::exception& e) {
+        return false;
+      }
+      return true;
+    }
+    candidate = candidate->parent();
+  }
+  return false;
+}
+
+int64_t estimateSerializedSize(const VectorPtr& vector) {
+  vector_size_t bytes = 0;
+  auto bytesPtr = &bytes;
+  VectorStreamGroup::estimateSerializedSize(
```

Comment:
It would be nice to add a version of estimateSerializedSize that takes a vector and returns the estimate in bytes for the whole vector:

```
auto bytes = VectorStreamGroup::estimateSerializedSize(vector);
```